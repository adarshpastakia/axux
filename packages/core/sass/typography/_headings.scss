// @author    : Adarsh Pastakia
// @version   : 0.0.1
// @copyright : 2021
// @license   : MIT

#{$prefix}-root {
  #{$prefix}-subheading {
    font-size: 0.75em;
    font-weight: $font-weight-normal;
  }

  @for $i from 1 through 6 {
    $heading: nth(nth($font-size-headings, $i), 1);
    $size: nth(nth($font-size-headings, $i), 2);

    #{$heading},
    #{$prefix}-heading--#{$i} {
      font-size: $size;
      font-weight: 300;

      small {
        display: block;
        @extend #{$prefix}-subheading;
      }
    }

    #{$prefix}-display--#{$i} {
      font-size: $size * 2.5;
      font-weight: 200;
    }
  }
}
