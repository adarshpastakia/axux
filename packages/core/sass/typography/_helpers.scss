// @author    : Adarsh Pastakia
// @version   : 0.0.1
// @copyright : 2021
// @license   : MIT

#{$prefix} {
  &-icon,
  &-avatar {
    display: inline-block;
    fill: currentColor;
    line-height: 1;
    position: relative;
    text-align: center;
    @include noselect();
    vertical-align: middle;

    > i,
    > img,
    > svg,
    > span {
      display: block;
      height: 1em;
      object-fit: contain;
      overflow: hidden;
      pointer-events: none;
      width: 1em;
    }

    &--round {
      > i,
      > span {
        font-size: 0.5em;
        height: 1.5em;
        padding: 0.25em;
        width: 1.5em;
      }
    }
  }

  &-icon {
    --spin-steps: 8;

    > span {
      font: inherit;
      text-align: center;

      > img {
        width: 100%;
      }
    }

    &--fluid {
      min-width: 1em;
      width: unset;
    }
    &--round {
      border-radius: 1em;
    }
    &--sm {
      font-size: 0.75em;
    }
    &--md {
      font-size: 1.25em;
    }
    &--lg {
      font-size: 1.5em;
    }
    &--xl {
      font-size: 2.5em;
    }
    &--xxl {
      font-size: 3.5em;
    }
  }

  @at-root [dir="rtl"] {
    .flippable #{$prefix}-icon,
    #{$prefix}-icon.flippable {
      svg,
      img {
        transform: scaleX(-1);
      }
    }
  }
  @at-root [dir="ltr"] {
    .flippable #{$prefix}-icon,
    #{$prefix}-icon.flippable {
      svg,
      img {
        transform: scaleX(1);
      }
    }
  }

  &-avatar {
    border-radius: 1em;
    font-size: 1.5em;
    height: 1.25em;
    line-height: 1.25em;
    padding: 0.125em;
    text-align: center;
    width: 1.25em;
    > span {
      display: inline-block;
      font-size: 0.5em;
      font-weight: $font-weight-medium;
      line-height: 1;
      vertical-align: 0.45em;
      width: unset;
    }
    > img {
      border-radius: 1em;
      overflow: hidden;
    }
    > #{$prefix}-icon {
      border-radius: 1em;
      overflow: hidden;
      vertical-align: 0.05em;
    }
    &--sm {
      font-size: 1.25em;
    }
    &--md {
      font-size: 2em;
    }
    &--lg {
      font-size: 2.5em;
    }
    &--xl {
      font-size: 3em;
    }

    &--topEnd,
    &--topStart,
    &--bottomStart,
    &--bottomEnd {
      box-shadow: 0 0 2px rgba(#000000, 0.5) inset;
      font-size: 0.5em;
      position: absolute;
    }
    &--topStart {
      left: 0;
      top: 0;
      transform: translate(-20%, -10%);
    }
    &--topEnd {
      right: 0;
      top: 0;
      transform: translate(20%, -10%);
    }
    &--bottomStart {
      bottom: 0;
      left: 0;
      transform: translate(-20%, 10%);
    }
    &--bottomEnd {
      bottom: 0;
      right: 0;
      transform: translate(20%, 10%);
    }
  }

  &-font {
    &--normal {
      font-size: $base-font-size;
    }
    &--inherit {
      font: {
        family: inherit;
        size: inherit;
        weight: inherit;
      }
    }

    &--reset {
      font-size: 1rem;
    }

    @each $label, $size in $font-sizes {
      &--#{$label} {
        font-size: $size;
      }
    }

    &--sans-serif {
      font-family: $font-family-sans-serif;
    }

    &--serif {
      font-family: $font-family-serif;
    }

    &--mono {
      font-family: $font-family-mono;
    }
  }

  &-text {
    vertical-align: middle;

    $clipClass: parse-dotless(#{$prefix}-text--clip);
    &[class^="#{$clipClass}"],
    &[class*=" #{$clipClass}"] {
      @include clip();
    }

    &--upper {
      text-transform: uppercase;
    }

    &--lower {
      text-transform: lowercase;
    }

    &--title {
      text-transform: capitalize;
    }

    &--smallcaps {
      font-variant: small-caps;
    }

    &--underline {
      text-decoration: underline;
    }

    &--strikethrough {
      text-decoration: line-through currentColor;
    }

    &--nowrap {
      white-space: nowrap;
    }

    &--break {
      word-break: break-word;
    }

    &--dropcaps {
      &::first-letter {
        float: left;
        font-size: 3em;
        font-weight: $font-weight-medium;
        line-height: 1;
        padding-right: 0.25em;
      }

      @at-root [dir="rtl"] &::first-letter {
        float: right;
        padding-left: 0.25em;
      }
    }

    > abbr {
      border-bottom: 1px dashed currentColor;
    }
  }

  // alignments
  &-align {
    &--start {
      text-align: start;
    }

    &--center {
      text-align: center;
    }

    &--end {
      text-align: end;
    }

    &--justify {
      text-align: justify;
    }

    &--middle {
      align-items: center;
      display: flex;
    }
  }

  // font weights
  &-weight {
    &--black {
      font-weight: $font-weight-black;
    }

    &--bold {
      font-weight: $font-weight-bold;
    }

    &--medium {
      font-weight: $font-weight-medium;
    }

    &--normal {
      font-weight: $font-weight-normal;
    }

    &--light {
      font-weight: $font-weight-light;
    }

    &--thin {
      font-weight: $font-weight-thin;
    }
  }

  &-link {
    cursor: pointer;
    display: inline-block;
    position: relative;

    &::before {
      @include transition(width 0.25s);
      bottom: 0;
      content: "";
      display: block;
      height: 1px;
      left: 0;
      margin: 0 auto;
      position: absolute;
      right: 0;
      width: 0;
    }

    &:hover::before {
      width: 100%;
    }
  }

  &-locale--link {
    align-items: center;
    cursor: pointer;
    display: flex;
    font-size: 1em;
    padding: 0.25em 0.5em;
    text-transform: capitalize;

    > span {
      flex: 1;
      padding: 0 1em 0 0.25em;
    }

    &::before {
      font-size: 1.1em;
    }

    &[data-selected="true"] {
      pointer-events: none;

      &::after {
        content: "\2713";
      }
    }

    &[data-locale="en"]::before {
      content: "\1F1FA\1F1F8";
    }
    &[data-locale="ar"]::before {
      content: "\1F1F8\1F1E6";
    }
    &[data-locale="de"]::before {
      content: "\1F1E9\1F1EA";
    }
    &[data-locale="es"]::before {
      content: "\1F1EA\1F1F8";
    }
    &[data-locale="fr"]::before {
      content: "\1F1EB\1F1F7";
    }
    &[data-locale="it"]::before {
      content: "\1F1EE\1F1F9";
    }
    &[data-locale="jp"]::before {
      content: "\1F1EF\1F1F5";
    }
    &[data-locale="ko"]::before {
      content: "\1F1F0\1F1F7";
    }
    &[data-locale="pt"]::before {
      content: "\1F1F5\1F1F9";
    }
    &[data-locale="ru"]::before {
      content: "\1F1F7\1F1FA";
    }
    &[data-locale="zh"]::before {
      content: "\1F1E8\1F1F3";
    }
  }
}

// indentation
@for $i from 1 through 4 {
  #{$prefix}-text--indent-#{$i} {
    text-indent: $base-font-size * $i;

    &#{$prefix}-text--dropcaps {
      text-indent: 0;

      &::first-letter {
        margin-inline-start: $base-font-size * $i;
      }
    }
  }

  #{$prefix}-text--clip-#{$i * 2} {
    --line-clamp: #{$i * 2};
  }
}
