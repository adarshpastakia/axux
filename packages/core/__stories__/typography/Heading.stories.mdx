import {
  Anchor,
  ArgsTable,
  Canvas,
  Description,
  Heading,
  Meta,
  Story,
  Title
} from "@storybook/addon-docs";
import { AxHeading } from "../../src";
import { ColorStory, DisplayStory, HeadingStory, SerifStory } from "./HeadingStory";

<Meta
  title="@core/Typography/Heading"
  component={AxHeading}
  parameters={{ jest: ["Heading.text.tsx"] }}
/>

<Anchor storyId="core-typography-heading--heading">
  <Title>Heading</Title>
</Anchor>

<Description of={AxHeading} />

<Canvas>
  <Story name="Heading">{HeadingStory}</Story>
</Canvas>

<Anchor storyId="core-typography-heading--display-heading">
  <Heading>Display Heading</Heading>
</Anchor>

<Canvas>
  <Story name="Display Heading">{DisplayStory}</Story>
</Canvas>

<Anchor storyId="core-typography-heading--serif-font">
  <Heading>Serif Font</Heading>
</Anchor>

<Canvas>
  <Story name="Serif Font">{SerifStory}</Story>
</Canvas>

<Anchor storyId="core-typography-heading--color">
  <Heading>Color</Heading>
</Anchor>

<Canvas>
  <Story name="Color">{ColorStory}</Story>
</Canvas>

<Heading>Props</Heading>

<ArgsTable of={AxHeading} sort="requiredFirst" />
