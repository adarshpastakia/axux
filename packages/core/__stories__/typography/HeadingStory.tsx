// @author    : Adarsh Pastakia
// @version   : 0.0.1
// @copyright : 2021
// @license   : MIT

import { Story } from "@storybook/react";
import { PropsWithChildren } from "react";
import { LIPSUM } from "../../../../storybook/components/Lipsum";
import { AxHeading } from "../../src";
import { HeadingProps } from "../../src/typography/Heading";

const Template: Story<PropsWithChildren<HeadingProps>> = (props) => <AxHeading {...props} />;

export const HeadingStory = Template.bind({});
HeadingStory.args = {
  children: LIPSUM.text,
  level: 1
};

export const DisplayStory = Template.bind({});
DisplayStory.args = {
  children: LIPSUM.text,
  level: 1,
  forDisplay: true
};

export const SerifStory = Template.bind({});
SerifStory.args = {
  children: LIPSUM.text,
  font: "serif"
};

export const ColorStory = Template.bind({});
ColorStory.args = {
  children: LIPSUM.text,
  bg: "lightest",
  color: "secondary",
  className: "ax-padding--sm"
};

export default { title: "Example/Heading", component: AxHeading };
