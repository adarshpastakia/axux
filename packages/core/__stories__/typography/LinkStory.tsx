// @author    : Adarsh Pastakia
// @version   : 0.0.1
// @copyright : 2021
// @license   : MIT

import { Story } from "@storybook/react";
import { PropsWithChildren } from "react";
import { BrowserRouter } from "react-router-dom";
import { AxLink } from "../../src";
import { LinkProps } from "../../src/typography/Link";

const Template: Story<PropsWithChildren<LinkProps>> = (props) => (
  <BrowserRouter>
    <AxLink {...props} />
  </BrowserRouter>
);

export const LinkStory = Template.bind({});
LinkStory.args = {
  children: "Simple link",
  href: "#/test"
};

export default { title: "Example/Link", component: AxLink };
