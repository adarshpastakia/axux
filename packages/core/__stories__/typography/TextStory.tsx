// @author    : Adarsh Pastakia
// @version   : 0.0.1
// @copyright : 2021
// @license   : MIT

import { Story } from "@storybook/react";
import { PropsWithChildren } from "react";
import { LIPSUM } from "../../../../storybook/components/Lipsum";
import { AxText } from "../../src";
import { TextProps } from "../../src/typography/Text";

const Template: Story<PropsWithChildren<TextProps>> = (props) => <AxText {...props} />;

export const TextStory = Template.bind({});
TextStory.args = {
  children: LIPSUM.para
};

export const SerifStory = Template.bind({});
SerifStory.args = {
  children: LIPSUM.para,
  className: "ax-text--dropcaps ax-align--justify",
  block: true,
  font: "serif"
};

export const ColorStory = Template.bind({});
ColorStory.args = {
  children: LIPSUM.para,
  className: "ax-text--indent-2 ax-align--justify ax-padding--sm",
  block: true,
  color: "secondary",
  bg: "lightest"
};

export default { title: "Example/Text", component: AxText };
