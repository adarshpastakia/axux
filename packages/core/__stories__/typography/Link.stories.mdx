import {
  Anchor,
  ArgsTable,
  Canvas,
  Description,
  Heading,
  Meta,
  Story,
  Title
} from "@storybook/addon-docs";
import { AxLink } from "../../src";
import { LinkStory } from "./LinkStory";

<Meta title="@core/typography/Link" component={AxLink} parameters={{ jest: ["Link.test.tsx"] }} />

<Anchor storyId="core-typography-link--link">
  <Title>Anchor Link</Title>
</Anchor>

<Description of={AxLink} />

<Canvas>
  <Story name="Link">{LinkStory}</Story>
</Canvas>

<Heading>Props</Heading>

<ArgsTable of={AxLink} sort="requiredFirst" />
