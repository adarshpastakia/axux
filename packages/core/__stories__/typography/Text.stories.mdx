import {
  Anchor,
  ArgsTable,
  Canvas,
  Description,
  Heading,
  Meta,
  Story,
  Title
} from "@storybook/addon-docs";
import { AxFlexBox, AxText } from "../../src";
import { ColorStory, SerifStory, TextStory } from "./TextStory";
import { LIPSUM } from "../../../../storybook/components/Lipsum";

<Meta title="@core/Typography/Text" component={AxText} parameters={{ jest: ["Text.test.tsx"] }} />

<Anchor storyId="core-typography-text--text">
  <Title>Text</Title>
</Anchor>

<Description of={AxText} />

<Canvas>
  <Story name="Text">{TextStory}</Story>
</Canvas>

<Anchor storyId="core-typography-text--serif-font">
  <Heading>Serif Font</Heading>
</Anchor>

<Canvas>
  <Story name="Serif Font">{SerifStory}</Story>
</Canvas>

<Anchor storyId="core-typography-text--color">
  <Heading>Color</Heading>
</Anchor>

<Canvas>
  <Story name="Color">{ColorStory}</Story>
</Canvas>

<Anchor storyId="core-typography-text--weights">
  <Heading>Weights</Heading>
</Anchor>

<Canvas columns={6}>
  <Story name="Weights">
    <AxFlexBox.Row>
      <AxText weight="black">Black text</AxText>
      <AxText weight="bold">Bold text</AxText>
      <AxText weight="medium">Medium text</AxText>
      <AxText weight="normal">Normal text</AxText>
      <AxText weight="light">Light text</AxText>
      <AxText weight="thin">Thin text</AxText>
    </AxFlexBox.Row>
  </Story>
</Canvas>

<Anchor storyId="core-typography-text--sizes">
  <Heading>Sizes</Heading>
</Anchor>

<Canvas columns={6}>
  <Story name="Sizes">
    <AxFlexBox.Row align="middle">
      <AxText size="xs">XSmall text</AxText>
      <AxText size="sm">Small text</AxText>
      <AxText size="default">Default text</AxText>
      <AxText size="md">Medium text</AxText>
      <AxText size="lg">Large text</AxText>
      <AxText size="xl">XLarge text</AxText>
      <AxText className="ax-col--full" size="2em">
        2em text
      </AxText>
    </AxFlexBox.Row>
  </Story>
</Canvas>

<Anchor storyId="core-typography-text--clipped">
  <Heading>Clipped</Heading>
</Anchor>

<Canvas>
  <Story name="Clipped">
    <div style={{ maxWidth: 1200 }}>
      <AxText clip={7}>
        <p>{LIPSUM.para}</p>
        <p>{LIPSUM.para}</p>
      </AxText>
    </div>
  </Story>
</Canvas>

<Anchor storyId="core-typography-text--marked-text">
  <Heading>Marked text</Heading>
</Anchor>

<Canvas>
  <Story name="Marked text">
    <AxText mark={LIPSUM.text}>{LIPSUM.para}</AxText>
  </Story>
</Canvas>

<Anchor storyId="core-typography-text--tooltip-text">
  <Heading>Tooltip text</Heading>
</Anchor>

<Canvas>
  <Story name="Tooltip text">
    <AxText tooltip="Test tooltip">{LIPSUM.text}</AxText>
    <AxText>{LIPSUM.para.replace(LIPSUM.text, "")}</AxText>
  </Story>
</Canvas>

<Heading>Props</Heading>

<ArgsTable of={AxText} sort="requiredFirst" />
