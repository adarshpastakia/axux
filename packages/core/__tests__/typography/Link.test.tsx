// @author    : Adarsh Pastakia
// @version   : 0.0.1
// @copyright : 2021
// @license   : MIT

import { composeStories } from "@storybook/testing-react";
import { fireEvent, render } from "@testing-library/react";
import * as stories from "../../__stories__/typography/LinkStory";

const { LinkStory } = composeStories(stories);

describe("Links", () => {
  it("should render link", (done) => {
    const fragment = render(<LinkStory />);
    expect(fragment.container).toHaveTextContent("Simple link");
    expect(fragment.container.querySelector(".ax-link")).toBeInTheDocument();
    expect(fragment.container).toMatchSnapshot();
    fragment.unmount();
    done();
  });

  it("should render new tab link", (done) => {
    const fragment = render(<LinkStory target="_blank" />);
    expect(fragment.container).toHaveTextContent("Simple link");
    expect(fragment.container.querySelector("[target='_blank']")).toBeInTheDocument();
    expect(fragment.container).toMatchSnapshot();
    fragment.unmount();
    done();
  });

  it("should handle onClick", (done) => {
    const handler = jest.fn();
    const fragment = render(<LinkStory className="test" onClick={handler} />);
    fireEvent.click(fragment.getByText("Simple link"));
    expect(handler).toBeCalled();
    fragment.unmount();
    done();
  });
});
