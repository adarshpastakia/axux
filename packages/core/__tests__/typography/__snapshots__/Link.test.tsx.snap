// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Links should render link 1`] = `
<div>
  <a
    aria-current="page"
    class="ax-link  active"
    href="/#/test"
    rel="noreferrer"
  >
    Simple link
  </a>
</div>
`;

exports[`Links should render new tab link 1`] = `
<div>
  <a
    aria-current="page"
    class="ax-link  active"
    href="/#/test"
    rel="noreferrer"
    target="_blank"
  >
    Simple link
  </a>
</div>
`;
